Making log_record async-safe

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
log_record is called, either directly or indirectly via log_event, from various signal handlers in various PBS daemons. log_record is not currently async-safe as it calls system calls like localtime_r() which are not async-safe. So, this PR is to make log_record async-safe by blocking all signals.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
Added code to block all signals inside log_record()

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
There's no good test for this, this solves a slim race condition, so code inspection might be the only way to analyze the changes + smoke tests to ensure that nothing breaks.


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

subhasisb commented:
Thanks, this was long pending!

mike0042 commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -947,6 +956,13 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
+#ifndef WIN32
+	sigprocmask(SIG_SETMASK, &old_mask, NULL);
+#else
+	return;
```
Shouldn't have to call return as the last line of a void function. Otherwise, looks good.

agrawalravi90 commented:
@mike0042 looks like Windows doesn't like an empty label, I'll need to revert the last commit

subhasisb commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -884,16 +893,16 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
 			snprintf(microsec_buf, sizeof(microsec_buf), ".%06ld", (long)tp.tv_usec);
 	}
 
+	/* lock the log mutex */
```
moving the mutex log before the call to localtime actually is not required. It will not make anything any better. If there is a fork that happens and takes the localtime_r's internal mutex to deadlock, this will not prevent it. The original place for the mutex is actually fine, i.e. it only guards against the switching of the log files

subhasisb commented on src/lib/Liblog/pbs_log.c:
```diff
@@ -908,7 +917,7 @@ log_record(int eventtype, int objclass, int sev, const char *objname, const char
 			fclose(logfile);
 		}
-		return;
+		goto sigunblock;
```
This jump will be bad, if it happens, will leave the mutex locked given where you have it now

agrawalravi90 replied:
We unlock the mutex at the beginning of this block (see line 913), so I think it should be fine.

agrawalravi90 commented:
@subhasisb I've addressed your comment, please let me know if it looks okay now. Also, any idea why where are 2 Travis builds on this PR? @hirenvadalia do you know?

hirenvadalia commented:
@agrawalravi90 Looks like some intermediate network issue, it's passing now. If it fails again let me know and don't restart it for debugging...
